presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -223,11 +224,22 @@ public PlanWithProperties visitOutput(OutputNode node, Context context)
         {
             PlanWithProperties child = planChild(node, context.withPreferredProperties(PreferredProperties.undistributed()));
 
+            // if the child is already a gathering exchange, don't add another exchange
+            if ((child.getNode() instanceof ExchangeNode) && ((ExchangeNode) child.getNode()).getType() == ExchangeNode.Type.GATHER) {
+                return rebaseAndDeriveProperties(node, child);
+            }
+
             if (!child.getProperties().isSingleNode()) {
                 child = withDerivedProperties(
                         gatheringExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
                         child.getProperties());
             }
+            else if (!child.getProperties().isCoordinatorOnly()) {
```

Comment:
When would be the case that `child.getProperties().isCoordinatorOnly()` is not true for output node? Such plan would be incorrect (since results would land on non-coordinator), right? I actually think that coordinator only distribution for `OutputNode` is set in `PlanFragmenter`. The distribution code is pretty tricky and interconnected.

However, in your case, I think you need to do completely different `if` in `visitOutput` (if you want to offload coordinator). You need to ensure that child plan node is an exchange. If it isn't then you should explicitly add gathering exchange.

I would change `if` here: https://github.com/prestodb/presto/pull/8074/commits/4d52a6fc8f1814951b3e46563379b1f9e7c0a99c#diff-3147362e4362824887f1eb890a167d41R232:
```
if (!child.getProperties().isSingleNode() || !(child instanceof ExchangeNode)) {
 ...
```